src/rust/engine/process_execution/docker/src/docker_tests.rs

```diff
@@ -106,54 +212,158 @@ async fn runner_errors_if_docker_image_not_set() {
+    fn new(image_name: &str) -> Self {
+        ExistingContainerTestRunner {
+            image_name: image_name.to_string(),
+            platform: platform_for_tests().unwrap(),
+            container_id: OnceLock::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl DockerCommandTestRunner for ExistingContainerTestRunner {
+    async fn setup<'a>(
+        &self,
+        store: Store,
+        executor: task_executor::Executor,
+        docker: &'a DockerOnceCell,
+        image_pull_cache: &'a ImagePullCache,
+        work_dir_base: PathBuf,
+        immutable_inputs: ImmutableInputs,
+    ) -> Result<crate::docker::CommandRunner<'a>, String> {
+        let command_runner = crate::docker::CommandRunner::new(
+            store,
+            executor,
+            docker,
+            image_pull_cache,
+            work_dir_base,
+            immutable_inputs,
+        )?;
+        match self.container_id.set(
+            command_runner
+                .container_cache
+                .container_for_image(&self.image_name, &self.platform, "")
+                .await?,
+        ) {
+            Ok(()) => Ok(command_runner),
+            Err(_) => Err("An error occurred when attempting to save container ID".to_string()),
+        }
+    }
+}
+
+struct MissingContainerTestRunner {
+    inner: ExistingContainerTestRunner,
+}
+
+impl MissingContainerTestRunner {
+    fn new(image_name: &str) -> Self {
+        MissingContainerTestRunner {
+            inner: ExistingContainerTestRunner::new(image_name),
+        }
+    }
+}
+
+#[async_trait]
+impl DockerCommandTestRunner for MissingContainerTestRunner {
+    async fn setup<'a>(
+        &self,
+        store: Store,
+        executor: task_executor::Executor,
+        docker: &'a DockerOnceCell,
+        image_pull_cache: &'a ImagePullCache,
+        work_dir_base: PathBuf,
+        immutable_inputs: ImmutableInputs,
+    ) -> Result<crate::docker::CommandRunner<'a>, String> {
+        let command_runner = self
+            .inner
+            .setup(
+                store,
+                executor,
+                docker,
+                image_pull_cache,
+                work_dir_base,
+                immutable_inputs,
+            )
+            .await?;
+        let container_id = self.inner.container_id.get().unwrap().0.as_str();
+        ContainerCache::remove_container(
+            docker.get().await?,
+            container_id,
+            Some(RemoveContainerOptions {
+                force: true,
+                ..RemoveContainerOptions::default()
+            }),
+        )
+        .await
+        .map_err(|err| {
+            format!("Failed to remove Docker container during missing container test setup `{container_id}`: {err:?}")
+        })?;
+        Ok(command_runner)
+    }
+}
+
+#[parameterized(runner = {&DefaultTestRunner, &ExistingContainerTestRunner::new(IMAGE), &MissingContainerTestRunner::new(IMAGE)}, name = {"default", "existing", "missing"})]
```

Comment:
Idle discussion: do you know of a way to reduce the repetition of many tests being parametrised with identical parameters?